### DISCUSSION TITLE
Macros for controlling instruments

### DISCUSSION THREAD

**Discussion by YulienPohl** [pymeasure/pymeasure#618] [Ideas]
Are there any plans to support macros to control multiple devices.
I have looked at [PyGMI](https://github.com/Argonne-National-Laboratory/PyGMI) which does somewhat do the trick, but is not under active delopment.
The current sequencer only allows to execute simple measurements.

For example considering a tube furnace.
1. Set Oven Temperature to 500 K @ 50 K/min
2. Wait 60 secs
3. Set Valve Flowrate to 500 mln/min
4. Wait for Oven Temperature to reach 500 K
5. Start Measurement
2. Wait 60 secs
6. Stop Measurements

**Comment by LongnoseRob** [comment]
@YulienPohl I like the idea, but on the other hand am also worried about something like this.
To stay with your Example:
If in this exammple the gas is N2 and the oven can only do 750C, nothing bad might happen if the PC running the control creashed and the gas-flow is not stopped.
But now imagine, the temperature is 1500C and the gas is something more reactive, a PC crash might cause some seriuos damage.
For certain security/safety related issues I would not trust a Python-script  running on some PC open available in a lab.

I can image, that for applciations where there are no issues when the measurement-"routine" is interrupted, this is an idea worth discussing.
Do you have some more ideas how to actually implment this?

There was a similiar disucssion in #611, maybe it would be a good idea to start from there?

**Reply by YulienPohl** [reply]
For implementation idea, I thought about to code it firstly separate from the widget, unlike the Sequencer which is intertwined with the gui.
For the widget, there are two possibilities:
1. using a text box which would require a parser for the keywords and values.
2. using a table with dropdown menus and value fields for each task.

For the detailed implementation of the backend, we could collect possible ideas from [bluesky](https://nsls-ii.github.io/bluesky/plans.html) or other libraries.

**Comment by msmttchr** [comment]
I would consider the option to use only python scripts with no GUI for such tasks. In such case, you would have the complete flexibility offered by python and use just the instrument part of Pymeasure. Have a look at https://github.com/pymeasure/pymeasure/blob/0400c37c9fa474afe8fbddb041de6f5bb708a259/examples/Basic/script.py for an example

**Reply by bilderbuchi** [reply]
This seems like the most straightforward approach.
